File: programmers/코딩테스트 고득점 Kit/스택, 큐/프린터.cpp
#include <string>
#include <vector>
#include <queue>
#include<functional>
using namespace std;
queue<pair<int, int>> q;
priority_queue<int, vector<int>, less<int>> pq;
int solution(vector<int> priorities, int location) {
	int answer = 0;
	for (int i = 0; i < priorities.size(); i++) {
		pq.push(priorities[i]);
		int val = 0;
		if (i == location) val = 1;
		q.push({ priorities[i], val });
	}
	while (!pq.empty()) {
		if (pq.top() == q.front().first) {
			answer++;
			pq.pop();
			if (q.front().second == 1) break;
		}
		else q.push(q.front());
		q.pop();
	}
	return answer;
}

File: programmers/코딩테스트 고득점 Kit/스택, 큐/프린터.java
import java.util.*;
class Document {
    private int location;
    private int priority;
    Document(int location, int priority) {
        this.location = location;
        this.priority = priority;
    }
    public int getLocation() {
        return location;
    }
    public int getPriority() {
        return priority;
    }
}
class Solution {
    public int solution(int[] priorities, int location) {
        int answer = 0;
        Queue<Integer> pQueue = new PriorityQueue<>(Comparator.reverseOrder());
        Queue<Document> queue = new LinkedList<>();
        for(int i=0;i<priorities.length;i++) pQueue.add(priorities[i]);
        for(int i=0;i<priorities.length;i++) queue.add(new Document(i, priorities[i]));
        int order = 1;
        while(!pQueue.isEmpty()) {
            if(queue.peek().getPriority() == pQueue.peek()) {
                if(queue.peek().getLocation() == location) {
                    answer = order;
                    break;
                }
                queue.remove();
                pQueue.remove();
                order++;
            }
            else {
                queue.add(queue.remove());
            }
        }
        return answer;
    }
}